src/TotalDissipationRateEquationSystem.C

```diff
@@ -0,0 +1,615 @@
+void
+TotalDissipationRateEquationSystem::initialize()
+{
+  solverAlgDriver_->initialize_connectivity();
+  linsys_->finalizeLinearSystem();
+}
+
+//--------------------------------------------------------------------------
+//-------- reinitialize_linear_system --------------------------------------
+//--------------------------------------------------------------------------
+void
+TotalDissipationRateEquationSystem::reinitialize_linear_system()
+{
+  // If this is decoupled overset simulation and the user has requested that the
+  // linear system be reused, then do nothing
+  if (decoupledOverset_ && linsys_->config().reuseLinSysIfPossible())
+    return;
+
+  // delete linsys
+  delete linsys_;
+
+  // create new solver
+  std::string solverName =
+    realm_.equationSystems_.get_solver_block_name("total_dissipation_rate");
+  LinearSolver* solver = realm_.root()->linearSolvers_->reinitialize_solver(
+    solverName, realm_.name(), EQ_TOT_DISS_RATE);
+  linsys_ = LinearSystem::create(realm_, 1, this, solver);
+
+  // initialize
+  solverAlgDriver_->initialize_connectivity();
+  linsys_->finalizeLinearSystem();
+}
+
+//--------------------------------------------------------------------------
+//-------- assemble_nodal_gradient() ---------------------------------------
+//--------------------------------------------------------------------------
+void
+TotalDissipationRateEquationSystem::assemble_nodal_gradient()
+{
+  const double timeA = -NaluEnv::self().nalu_time();
+  nodalGradAlgDriver_.execute();
+  timerMisc_ += (NaluEnv::self().nalu_time() + timeA);
+}
+
+//--------------------------------------------------------------------------
+//-------- compute_effective_flux_coeff() ----------------------------------
+//--------------------------------------------------------------------------
+void
+TotalDissipationRateEquationSystem::compute_effective_diff_flux_coeff()
+{
+  const double timeA = -NaluEnv::self().nalu_time();
+  effDiffFluxAlg_->execute();
+  timerMisc_ += (NaluEnv::self().nalu_time() + timeA);
+}
+
+//--------------------------------------------------------------------------
+//-------- predict_state() -------------------------------------------------
+//--------------------------------------------------------------------------
+void
+TotalDissipationRateEquationSystem::predict_state()
+{
+  const auto& ngpMesh = realm_.ngp_mesh();
+  const auto& fieldMgr = realm_.ngp_field_manager();
+  const auto& tdrN = fieldMgr.get_field<double>(
+    tdr_->field_of_state(stk::mesh::StateN).mesh_meta_data_ordinal());
+  auto& tdrNp1 = fieldMgr.get_field<double>(
+    tdr_->field_of_state(stk::mesh::StateNP1).mesh_meta_data_ordinal());
+
+  const auto& meta = realm_.meta_data();
+  const stk::mesh::Selector sel =
+    (meta.locally_owned_part() | meta.globally_shared_part() |
+     meta.aura_part()) &
+    stk::mesh::selectField(*tdr_);
+  nalu_ngp::field_copy(ngpMesh, sel, tdrNp1, tdrN);
+  tdrNp1.modify_on_device();
```

Comment:
FYI the `modify_on_device` flag is set inside `field_copy` so this is unnecessary. It is a null op though so it doesn't matter. Might be worth adding a sync call on `tdrN` though. @alanw0 is there a reason we don't do a sync on the src field inside `copy_field` too?